### ISSUE TITLE
CID V1 with rawLeaves option for Mutable File System Incorrect?

### ISSUE THREAD

**Issue by skogard** [ipfs/js-ipfs#3745]
### Version

```
{
  version: '0.8.0',
  repo: '10',
  commit: '81f944163f5a78ef15fdc23b2c488ba7f44af0be',
  'interface-ipfs-core': '^0.147.0'
}
```

### Platform

Linux/Mac/etc.

### Subsystem

IPFS MFS (Mutable File System) https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-core/src/components/files

### Severity:

High: IPFS CID does not work as expected when using `rawLeaves: true`

### Description:

What you did

I tried a simple `ipfs.files.write()`, but with `cidVersion: 1` and `rawLeaves: true`. Here's the code:

```
const Ipfs = require('ipfs-core');
(async () => {
  const ipfs = await Ipfs.create()
  await ipfs.files.write('/hello.txt', Buffer.from('Hello world'), {
    create: true,
    cidVersion: 1,
    rawLeaves: true
  });
  const directoryStatus = await ipfs.files.stat('/')
  console.log(directoryStatus.cid.toString())
  const result = []
  for await (const resultPart of ipfs.files.ls('/')) {
    result.push(resultPart)
  }
  console.log("LS", result)
})();
```

What happened: Instead of returning a `bafk...` address it returns `bafy...` address.

### What you expected to happen

I expected the leaf node (in this case `hello.txt`) to be of `raw` type and not `dag-pb` type, which would have a `bafk...` address, which is supposed to be the version for when `rawLeaves: true` is set.

### What actually happened

As can be seen from below, the `hello.txt` has the CID of `bafybeigqrzobiyfhsnhoh2d3kxmmwv5z76od7ty4alhywz43umkgnlw7rm`, which is a hash that uses dag-pb (not raw) (See https://cid.ipfs.io/#bafybeigqrzobiyfhsnhoh2d3kxmmwv5z76od7ty4alhywz43umkgnlw7rm)

```
$ node index
Swarm listening on /ip4/127.0.0.1/tcp/4002/p2p/QmRS8Z6YopPVqci2Zq2AdaysWifsUwd6gPwV2hCXdVhFYF
Swarm listening on /ip4/192.168.1.48/tcp/4002/p2p/QmRS8Z6YopPVqci2Zq2AdaysWifsUwd6gPwV2hCXdVhFYF
Swarm listening on /ip4/169.254.165.142/tcp/4002/p2p/QmRS8Z6YopPVqci2Zq2AdaysWifsUwd6gPwV2hCXdVhFYF
Swarm listening on /ip4/127.0.0.1/tcp/4003/ws/p2p/QmRS8Z6YopPVqci2Zq2AdaysWifsUwd6gPwV2hCXdVhFYF
bafybeig2tjq75tuuiftxcjvlopnfx5szhxdogsyqdovrci2u3jyrxo5uqm
LS [
  {
    cid: CID(bafybeigqrzobiyfhsnhoh2d3kxmmwv5z76od7ty4alhywz43umkgnlw7rm),
    name: 'hello.txt',
    type: 'file',
    size: 11,
    mode: 420,
    mtime: undefined
  }
]
```

**Comment by achingbrain** [comment]
The MFS uses [trickle dag](https://discuss.ipfs.io/t/what-is-the-difference-between-trickle-dag-and-merkle-dag/265/2) structures, the root of which is always a dag-pb node.  If the data you are adding is big enough to need leaf nodes to represent the dag, they will be raw nodes when the `rawLeaves` option is passed true.

This behaviour has been inherited from the go-ipfs implementation of the MFS.

If you want to add a raw node, either use `ipfs.add(buf, { rawLeaves: true })` or `ipfs.dag.put(buf, { format: 'raw' })`.  If you then want that raw node in your mfs, use `ipfs.files.cp(cid, '/some-path.tx')` where `cid` is the output of `ipfs.add` or `ipfs.dag.put` above to do that.

**Comment by skogard** [comment]
Thank you for the answer, I managed to get the CID to match by following your instructions.

But I don't know if this actually resolves the problem fundamentally. Or maybe I misunderstand. Please let me know if I am mistaken.

Basically if I understand correctly, if I want to make sure that the MFS behavior matches the go-ipfs implementation (which uses `rawLeaves: true` by default), it's not as simple as using the MFS API. Instead I have to go through the process of using `ipfs.add()` or `ipfs.dag.put()` and then `ipfs.files.cp()` it into the tree. Is this correct?

If this is the case, my issue is, doesn't this mean we can't use IPFS-JS based MFS if we want it to match the behavior of go-ipfs? This means the same folder added using JS version of MFS will have a different root CID than the version added using go-ipfs. (This is actually what's happening in my case, which is why I opened the issue)

**Comment by achingbrain** [comment]
> I want to make sure that the MFS behavior matches the go-ipfs implementation (which uses rawLeaves: true by default)

go-ipfs does not use `rawLeaves: true` by default:

```console
$ echo 'hello' | ipfs files write /file-1.txt --create
$ ipfs files ls -l /file-1.txt
file-1.txt	QmcFUHo1DBpydnTSeuXszKgQqnXQCJq6Wu1BUdzEfuRahM	6   // <- dag-pb
$ ipfs ls QmcFUHo1DBpydnTSeuXszKgQqnXQCJq6Wu1BUdzEfuRahM
QmQnPyGFkJ6NufTCS26nTycwenbp2XGBZj8BuNaXs9zgD4 6  // <-  leaf, also dag-pb
```

If you use the [webui](https://webui.ipfs.io/) to inspect that leaf node CID, you'll see that it's a [UnixFS node of type "raw"](https://github.com/ipfs/specs/blob/master/UNIXFS.md#data-format) - but it's still a UnixFS node, that is, it's a structure wrapped in a protobuf.

![image](https://user-images.githubusercontent.com/665810/126031784-2b39b566-192a-4828-b578-d00a2348d06d.png)

You have to explicitly specify raw leaves:

```console
$ echo 'hello' | ipfs files write /file-2.txt --create --raw-leaves
$ ipfs files ls -l /file-2.txt
file-2.txt	QmcHRf1w1NG9Ld9qUWAPyPKmCsESeHvu3kaa7By4pBvSgw	6   // <- dag-pb
$ ipfs ls QmcHRf1w1NG9Ld9qUWAPyPKmCsESeHvu3kaa7By4pBvSgw
bafkreicysg23kiwv34eg2d7qweipxwosdo2py4ldv42nbauguluen5v6am 6    // <- leaf, raw node
```

The `bafk` CID here is a v1 CID with the raw codec, that basically says 'interpret these bytes as just bytes'.  This wouldn't work with the `Qm..` leaf node in the first example because the bytes the CID refers to are inside a protobuf wrapper.

> This means the same folder added using JS version of MFS will have a different root CID than the version added using go-ipfs. (This is actually what's happening in my case, which is why I thought this is a bug and opened the issue)

That does sound like a bug, the same data with the same args should result in the same CID across implementations.

If you have a repro case would you like to open a PR to the unixfs importer that fixes the issue?  The trickle dag generator is implemented in js [here](https://github.com/ipfs/js-ipfs-unixfs/blob/master/packages/ipfs-unixfs-importer/src/dag-builder/file/trickle.js) and in go [here](https://github.com/ipfs/go-unixfs/blob/master/importer/trickle/trickledag.go), it's possible there's a mismatch between them.

**Comment by lidel** [comment]
>  If this is the case, my issue is, doesn't this mean we can't use IPFS-JS based MFS if we want it to match the behavior of go-ipfs? This means the same folder added using JS version of MFS will have a different root CID than the version added using go-ipfs. (This is actually what's happening in my case, which is why I thought this is a bug and opened the issue)

Shouldn't the MFS API automatically handle this so that it matches the behavior of go-ipfs?

@skogard iiuc this is not a difference between js-ipfs and go-ipfs, but between default settings used by `ipfs add` and `ipfs files write` in both JS and GO.

TLDR is that  `ipfs files write --create` is equivalent of `ipfs add --trickle && ipfs files cp`  and by default, `ipfs add` does not use `--trickle` that is why you get different CIDs. This is by design, different DAG types [serve different purposes](https://github.com/ipfs-inactive/support/issues/45#issuecomment-269296211).

if you want  CIDv1 produced by `ipfs add --cid-version 1` you need to use `ipfs add  --cid-version 1 + files cp` operation as noted by @achingbrain 

You can read a longer explainer at:
- ["ipfs add" and "ipfs files write" commands returns different hashes](https://github.com/ipfs-inactive/support/issues/45)
- [Upload via WebUI (ipfs.files.write --create) produces different hash than CLI (ipfs.add)](https://github.com/ipfs/ipfs-webui/issues/676)
